### ISSUE TITLE
Concrete type only used in `impl Trait` is unused

### ISSUE THREAD

**Issue by cramertj** [rust-lang/rust#55124]
The following code gives "warning: enum is never used: `Empty`":

```rust
enum Empty { }
trait Bar<T> {}
impl Bar<Empty> for () {}

fn boo() -> impl Bar<Empty> {}

fn main() {
    boo();
}
```

This regressed pretty recently. cc @oli-obk

**Comment by jonas-schievink** [comment]
1.27.0 does not warn, 1.28.0 does.

**Comment by pnkfelix** [comment]
triage: P-medium; This seems like a papercut that should be easy to work-around (at least if one isn't using `forbid`...).

**Comment by pnkfelix** [comment]
switched to P-high after prompting from @nikomatsakis

**Comment by pnkfelix** [comment]
whoops forgot to tag with T-compiler in time for the meeting...

**Comment by pnkfelix** [comment]
assigning to @nikomatsakis

**Comment by oli-obk** [comment]
I'm grabbing this, I think I broke it and it should be easy to fix
